Sources/kai/Parser.swift

```diff
@@ -342,6 +355,87 @@ extension Parser {
                 return AstNode.invalid(location ..< body.endLocation)
             }
 
+        case .keyword(.switch):
+            let (_, location) = try consume(.keyword(.switch))
+            
+            let cond: AstNode?
+            switch try lexer.peek()?.kind {
+            case .lbrace?:
+                cond = nil
+                
+            default:
+                // Make sure an identifier doesn't get turned into a
+                // compound literal
+                let prevState = state
+                defer { state = prevState }
+                state.insert(.disallowCompoundLiteral)
+                
+                cond = try expression()
+            }
+            
+            try consume(.lbrace)
+            
+            var cases: [AstNode] = []
+            var defaultBody: AstNode? = nil
+            
+            while try lexer.peek()?.kind != .rbrace {
+                try consumeTerminators(justNewlines: true)
+                
+                switch try lexer.peek()?.kind {
```

Comment:
The more _typical_ way to do this currently would be to have _sub-parsers_ for `case` and for `default` this way each parser is a smaller component.

For example you could refactor the loop body to:
```swift
while try lexer.peek()?.kind != .rbrace {
    
    try consumeTerminators(justNewlines: true)
    let caseExpr = try expression()
    guard case .stmtCase = caseExpr else {
        report("Expected switch case got \(caseExpr)", caseExpr)
        continue
    }
}
try consume(.rbrace)
```

for the case (and default sub-parsers)
```swift

case .keyword(.case): // as a `nud` will also need something for a `isDefault` possibly on the AstNode
    try consume(.keyword(.case))
    let (_, colon) = try consume(.colon)
    guard state.contains(.permitCaseOrDefault) else {
        reportError("syntax error, unexpected case keyword outside of `switch`", colon)
        return .invalid(colon ..< colon)
    }
    let colonBp = try! lbp(.colon)
    let match = try expression(colonBp)
    
    var stmts: [AstNode] = []
    while let nextToken = try lexer.peek()?.kind, nextToken != .case, nextToken != .default {
        let stmt = try expression()
        try consumeTerminators()
        stmts.append(stmt)
    }

    let block = AstNode.stmtBlock(stmts, colon ..< (stmts.last ?? colon))

    // again, we maybe will want an `isDefault` here
    return AstNode.stmtCase(match, stmts,  colon ..< (stmts.last ?? colon))
```